### ISSUE TITLE
Erweiterung der Lambda Aufgabe um Comparator und Comparable

### ISSUE THREAD

**Issue by CrappyAlgorithm** [Programmiermethoden/Dungeon#407]
In der Aufgabe zu Lambda habe ich ja schon weiterführende Aufgaben zu Comparator und Comparable angelegt. Damit die nicht verloren gehen, notiere  ich die erstmal in diesem Issue. Wir müssen ja noch festlegen, wie wir das umsetzen. Da es ja eigentlich nur eine Erweiterung der Lambda Aufgabe ist wäre es irgendwie auch doof, das als separate Aufgabe mit einer Kopie der meisten Projektteile umzusetzen. 

**Aufgabentext:**

5. Machen Sie sich mit dem Interface [java.util.Comparator](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Comparator.html) vertraut.

    5.1) Erstellen Sie mithilfe eines Lambda-Ausdrucks ein Comparator-Objekt, um Student-Objekte nach ihrem Alter zu sortieren. Verwenden Sie den Comparator, um die Studierendenliste absteigend zu sortieren.

    5.2) Machen sie sich nun mit der Methode [java.util.Comparator#comparing](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Comparator.html#comparing(java.util.function.Function)) vertraut. Sie ermöglicht das Erstellen eines Comparators, durch die Übergabe eines Getters des Feldes, nachdem sortiert werden soll. Erstellen Sie einen Comparator, um die Studierendenliste aufsteigend nach dem Namen zu sortieren. _Hinweis_: Verwenden Sie dabei eine Methodenreferenz und keinen Lambda-Ausdruck.

6. Machen Sie sich mit dem Interface [java.util.Comparable](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html) vertraut. Erweitern Sie die Klasse Student um das Compareable-Interface zum Vergleichen über das Alter und implementieren Sie alle dafür nötigen Methoden. Sortieren Sie jetzt die Studierendenliste aufsteigend, ohne einen eigenen Comparator oder Lambda-Ausdruck zu verwenden.

**Lösung:**

5.1. 
```java
/**
     * 5.1: Sortiert die Studierendenliste aufsteigend nach dem Alter unter Verwendung eines Comparators.
     * Der Comparator soll mithilfe eines Lambda-Ausdrucks erzeugt werden.
     *
     * @param students die zu sortierende Liste der Studierenden
     * @return die sortierte Liste mit Studierenden
     */
    public static List<Student> sort_5a(List<Student> students) {
        Comparator<Student> comparatorByAge = (Student a, Student b) -> b.getBirthday().compareTo(a.getBirthday());
        students.sort(comparatorByAge);
        return students;
    }
```

5.2. 
```java
/**
     * 5.2: Sortiert die Studierendenliste aufsteigend nach dem Namen unter Verwendung eines Comparators.
     * Der Comparator soll mithilfe der comparing Methode und einer Methodenreferenz erzeugt werden.
     *
     * @param students die zu sortierende Liste der Studierenden
     * @return die sortierte Liste mit Studierenden
     */
    public static List<Student> sort_5b(List<Student> students) {
        Comparator<Student> comparatorByName = Comparator.comparing(Student::getName);
        students.sort(comparatorByName);
        return students;
    }
```

6.
```java
/**
     * 6: Sortiert die Studierendenliste aufsteigend nach Alter unter Verwendung des Comparable Interfaces.
     *
     * @param students die zu sortierende Liste der Studierenden
     * @return die sortierte Liste mit Studierenden
     */
    public static List<Student> sort_6(List<Student> students) {
        students.sort(Comparator.naturalOrder()); // sort(null) ist ebenfalls möglich
        return students;
    }
```

**Comment by AMatutat** [comment]
@cagix willst das behalten?

**Comment by cagix** [comment]
@AMatutat Ja, lass mal offen. Kriegen wir das irgendwie den Aufgaben zugeordnet? Sollte das doch besser ins Lecture-Repo? Im Grunde betrifft das "Aufgaben" und "D2A".

**Comment by AMatutat** [comment]
Lecture-Repo klingt irgendwie sinnvoll. 
Geht ja wirklich um Inhalt nicht um Strukturen, deswegen sehe ich das D2A Label auch hier nicht wirklich.
